### ISSUE TITLE
False positive for Mapping argument passed to Maplike | Iterable[tuple[...]] constructor parameter

### ISSUE THREAD

**Issue by jab** [astral-sh/ty#4278]
### Summary

ty incorrectly rejects a constructor call that passes a `Mapping[KT, VT]` argument to a parameter typed as `Maplike[KT, VT] | Iterable[tuple[KT, VT]]`, even though `Mapping` is structurally compatible with `Maplike`.

This surfaced in the bidict codebase when upgrading ty from `0.0.58` to `0.0.72`.

### Reproduction

```python
from __future__ import annotations

import typing as t
from collections.abc import Iterable, Mapping

KT = t.TypeVar("KT")
VT = t.TypeVar("VT")
VT_co = t.TypeVar("VT_co", covariant=True)

@t.runtime_checkable
class Maplike(t.Protocol[KT, VT_co]):
    def keys(self) -> Iterable[KT]: ...
    def __getitem__(self, key: KT, /) -> VT_co: ...

MapOrItems: t.TypeAlias = Maplike[KT, VT] | Iterable[tuple[KT, VT]]

class C(t.Generic[KT, VT]):
    def __init__(self, arg: MapOrItems[KT, VT] = (), /) -> None:
        pass

    def f(self, other: Mapping[KT, VT]) -> t.Self:
        new = self.__class__(other)
        return new
```

### Expected behavior

The constructor call should be accepted. `other` has type `Mapping[KT, VT]`, and `Mapping` provides `keys()` and `__getitem__()` with the right structural shape for `Maplike[KT, VT]`.

### Actual behavior

`ty 0.0.72` reports:

```text
error[invalid-argument-type]: Argument to `C.__init__` is incorrect
  --> /tmp/ty_ctor_repro.py:22:15
   |
22 |         new = self.__class__(other)
   |               ^^^^^^^^^^^^^^^^^^^^^ Expected `C[tuple[tuple[Never, Unknown], Unknown], Unknown]`, found `C[tuple[Never, Unknown], Unknown]`
```

### Regression window

This minimal reproducer:

- passes on `ty 0.0.58`
- passes on `ty 0.0.69`
- passes on `ty 0.0.70`
- passes on `ty 0.0.71`
- fails on `ty 0.0.72`

So this appears to be a regression introduced between `0.0.71` and `0.0.72`.

### Analysis

This looks like ty is mis-inferring the `MapOrItems[KT, VT]` constructor parameter when matching the `Mapping[KT, VT]` argument against the `Iterable[tuple[KT, VT]]` side of the union, instead of accepting it through the structurally compatible `Maplike[KT, VT]` side.

This affects the bidict codebase in `BidictBase.__ror__`, where the equivalent constructor call is currently suppressed as a false positive.

### Version

ty 0.0.72

**Comment by carljm** [comment]
Thanks for the report, and sorry for the regression! The immediate triggering change was https://github.com/astral-sh/ruff/pull/27580. I noticed the regression in that PR, but accepted it because the constraint-solver bug is pre-existing, unrelated, and already prioritized for fixing. Parallel examples with `list`, `set`, `frozenset` etc already failed before that PR; collapsing `tuple[Never]` to `Never` just happened to hide it for tuples.

This is indeed a bug in union matching in the generic constraint solver, as you correctly identified. A minimal example is just this:

```py
class C[T]:
    def __init__(self, value: T | tuple[T]) -> None:
        pass

    def copy(self, value: T) -> None:
        self.__class__(value)
```

The problem is that the bad union matching generates a self-referential `tuple[T] <: T` constraint, and the cycle-breaking in the constraint solver injects the `Never`.

The right fix here is in union constraint solving, and is already broadly described by #3557, but I'll keep this open as a sub-issue.
